src/Mod/Sketcher/App/planegcs/GCS.cpp

```diff
@@ -1326,13 +1326,63 @@ int System::addConstraintInternalAlignmentParabolaFocus(Parabola &e, Point &p1,
     return addConstraintEqual(e.focus1.y, p1.y, tagId, driving, Constraint::Alignment::InternalAlignment);
 }
 
-int System::addConstraintInternalAlignmentBSplineControlPoint(BSpline &b, Circle &c, int poleindex, int tagId, bool driving)
+int System::addConstraintInternalAlignmentBSplineControlPoint(BSpline &b, Circle &c, unsigned int poleindex, int tagId, bool driving)
 {
     addConstraintEqual(b.poles[poleindex].x, c.center.x, tagId, driving, Constraint::Alignment::InternalAlignment);
     addConstraintEqual(b.poles[poleindex].y, c.center.y, tagId, driving, Constraint::Alignment::InternalAlignment);
     return addConstraintEqual(b.weights[poleindex], c.rad, tagId, driving, Constraint::Alignment::InternalAlignment);
 }
 
+int System::addConstraintInternalAlignmentKnotPoint(BSpline &b, Point &p, unsigned int knotindex, int tagId, bool driving)
+{
+    if (b.periodic && knotindex == 0) {
+        // This is done here since knotpoints themselves aren't stored
```

Comment:
I think this was alluding to why the constraint isn't applied somewhere in `Sketch.cpp` or `planegcs/Constraints.h`. Most likely the former. See https://github.com/FreeCAD/FreeCAD/commit/919572210af90ab6066f543e38cf2aaa2553bcbf (the relevant commit before squash). There I (probably) wanted to add the coincidence constraint at time of `addBSpline`.

I won't be opposed to re-wording the comment to tell why we do this. There is a small efficiency hit we take by keeping four points in the solver (first and last knots and start and end points) which are all coincident, but that makes dragging etc hell because the knot is which needs to be "dragged". There might also be some merit to using a 0 tag as well, from the new knowledge I have.

Reply:
I understand.

I see nothing untoward in creating these solver constraints here. Many sketcher constraints end up implemented as several solver constraints, see for example:
https://github.com/FreeCAD/FreeCAD/blob/9e9c78404123797d93de47bbde28a78042565274/src/Mod/Sketcher/App/planegcs/GCS.cpp#L975

As I see it, we are adding an internal alignment constraint to a knot (first knot). When this knot meets some conditions, it has an additional special treatment. 

I think rewording the comment so that it can easier understood is the way to go.